### DISCUSSION TITLE
Multiple problems while making this project

### DISCUSSION THREAD

**Discussion by jumperrg** [alf45tar/PedalinoMini#322] [Q&A] [answered]
Hello! 

I want to say - thank you for this project, there is a really good ideas standing behind the project such as sequences, web interface, and multiple protocols.

But there are a lot of troubles that I found while building it. And it drives me crazy almost every time I'm making each part.

First of all - why do have you chosen Platformio?  It's laggy and makes additional complications. Why didn't you used just VS code with Arduino? It's easy to setup. But this is the smallest problem.

Here ara several problems and questions which I faced while building the project:

1. I want to use UART2 (pins 16, 17) as a MIDI port, where can I chose them? I've found  ESPSerialMIDI.h  with theese pins setup, but it haven't work for me, I see only high level on Tx pin, when I push the button (I've turned on Legacy midi and chose "program change" on button press in the web interface).

3. If I want to add another hardware, like display to each button(i want to use st7735 displays) where can I add my code for driving theese displays? Is there a certain approach? 

5. If I want to setup pins I should go to different files in order to find pins setup. Maybe it is a good idea to make a separate file like PinSetup.h to keet all the pins setup in one place? 

7. Web interface a little bit uncomfortable to use, when I have a long list of pedals and want to change each pedal behavior, which is quite frequent procedure, I should scroll down each time to press "Apply" and "Save" button. Maybe it's a godd idea to keep theese buttons always visible, like a footer.

P.S. I would like to join and contribute to this project especially with theese ideas, I believe this project has huge potential, but it lacks simplicity for the not high skilled people.

**Comment by alf45tar** [comment]
Let's try to reply your questions one by one. It will take more days.

**Comment by alf45tar** [comment]
I do not have any experience with VS Code and Arduino but it looks pretty basic (at least 4 years ago when I started).
PlatformIO looks not so bad for me, it is robust, fully featured, tons of boards and it has 2 times developers than Arduino extension.

**Comment by alf45tar** [comment]
1. Do not touch ESPSerialMIDI.h to change UART pins assignment. You must modify Pedalino.h instead.

https://github.com/alf45tar/PedalinoMini/blob/b4ec16f77268ff03b3feb71263d918024f5028a3/src/Pedalino.h#L87
https://github.com/alf45tar/PedalinoMini/blob/b4ec16f77268ff03b3feb71263d918024f5028a3/src/Pedalino.h#L88
https://github.com/alf45tar/PedalinoMini/blob/b4ec16f77268ff03b3feb71263d918024f5028a3/src/Pedalino.h#L89
https://github.com/alf45tar/PedalinoMini/blob/b4ec16f77268ff03b3feb71263d918024f5028a3/src/Pedalino.h#L90

**Comment by alf45tar** [comment]
2. Add one display for each button requires more knowledge than build PedalinoMini from my buggy documentation.

- Consider that memory is not unlimited and we are using 99% of it. You have to disable some features to create space.
- The approach for a new display is to create a copy of DisplayOLED.h and reimplement the functions you will find inside. This is not include the redesign of the graphics of course or increase the number of displays.

**Comment by alf45tar** [comment, marked as answer]
3. PinSetup.h is not a bad idea but it is not to hard to find them in Pedalino.h too

**Comment by alf45tar** [comment]
4. I am an expert of nothing. The web UI is the best I can do with my current knowledge, capabilities and time. But believe me it was very challenging to serve so huge pages with the limited resource of ESP32. I spent a lot of time to create something stable. Contributions from experts are welcome.

**Comment by alf45tar** [comment]
Last but not least PedalinoMini is fully configurable from hardware and software point of view. I am not providing a fixed configuration but everyone must configure his pedal. It is not for lazy people but for people want to learn somenthing during building something.
 
If you are looking for simplicity there are tons of MIDI pedals on GitHub with 4 buttons sending 4 fixed CC with around 200 lines of code.

For not high skilled people I recently introduced pre-compiled firmware and user friendly tool in the browser to

- Install & update firmware
- Connect device to the Wi-Fi network
- Visit the device's hosted web interface
- Access logs and send terminal commands

**Comment by jumperrg** [comment]
Thank you for the replies, and I'm sorry for my emotional first post. After I posted it I've found some answers. If I get successful with my ideas about web interfase and displays I will post my results.

**Comment by jumperrg** [comment]
Hi alf45tar! 
I want to ask you one more question, it's about expression pedal connection. 

I have M-audio expression pedal, and when I tried to connect it to the Pedalino - it worked but not as expected. I want to send control change signals with values from 0 to 127, but it sends something like between 0 and 5, and then instantly about 127. So it seems like ADC pin (I use D4 and A4) isn't properly configured because it's always 3.3 volts on it when exp pedal even not connected. 

So do I need to configure ADC pin additionally, or it is sufficient to choose pedal mode in the web interface? 

Thanks

**Comment by alf45tar** [comment]
D4 is maintained to 3.3V to provide power to pedal.
A4 is maintained at 3.3V by the pull-up resistor when not connected.
I do not see any problem here.

Check the web ui configuration.

Mode must be Analog.
Analog Response = Linear as starting point is ok. Try the other options later.
Min = 0
Max = 1023
Easing = 2 or 3
Activity Threshold = 8 or 16
Analog Calibration ON

**Reply by jumperrg** [reply]
Thank you, manual calibration helped a lot, I've set Min=450 and max - 1023.

**Reply by alf45tar** [reply]
> Thank you, manual calibration helped a lot, I've set Min=450 and max - 1023.

Does automatic calibration not work?

**Reply by jumperrg** [reply]
It worked in some way, but the automatic range keeped set wrongly  several times.

**Comment by jumperrg** [comment]
Hi alf45tar!
I can't understand, what does Color2  do?  

For example I want to do the following: 

1. Send Control Change command with value 0 on the 1st click and switch off the LED.
2. Send Control Change command with value 127 on the 2nd click and glow the LED red.

But it works not as intended. Doesn't matter what I choose on Color 1 - it starts to glow red after short delay, and sends 2 CC messages one after the other with value 0 and 127. 

But I don't need to send 2 messages on one click. How to send a single message? 
And what is slot component means? 

![image](https://user-images.githubusercontent.com/20793480/168479844-75b3dd4c-5ee0-4749-8789-f5864f22585e.png)

Thanks in advance!

**Reply by jumperrg** [reply]
I'm sorry, perhaps it was something confusing for me because I edited the preferences multiple times and something went wrong. I erazed memory and started from blank settings, and everything works properly now. So my last questions are not relevant.

**Reply by alf45tar** [reply]
Slot parameter is not used yet.

**Comment by jumperrg** [comment]
Another question - can I send Note on with certain velocity? Velocity parameters are greyed out 
![image](https://user-images.githubusercontent.com/20793480/168490383-2c863252-2035-4358-9d8f-1bb1976d5d77.png)

**Reply by alf45tar** [reply]
Yes, you can. As of today It has been fixed in sponsor version only.

**Comment by jumperrg** [comment]
Hi alf45tar!  
I have another question/bug report, don't know exacly.  Here is the video:  https://youtu.be/osStY3Eor3w

The problem is that when I set "#define PEDALS 8" in Pedalino.h  file - in the web interface I see that number of pedals not 8 but goes up to 103 !!!  

Another problem you can see at the video is that only 6 LEDs flashes on boot (sometimes 8 sometimes 6 )

https://youtu.be/osStY3Eor3w?t=115   -   8 leds on boot 

https://youtu.be/osStY3Eor3w?t=147   -   6 leds on boot

Sorry for video quality.
